kernel/aster-nix/src/net/socket/ip/datagram/mod.rs

```diff
@@ -284,6 +284,68 @@ impl Socket for DatagramSocket {
         poll_ifaces();
         Ok(sent_bytes)
     }
+
+    fn sendmsg(&self, msg_hdr: MessageHeader, flags: SendRecvFlags) -> Result<usize> {
+        let MessageHeader {
+            addr,
+            io_vec_iter,
+            control_message,
+        } = msg_hdr;
+
+        if control_message.is_some() {
+            // TODO: support sending control message
+            warn!("sending control message is not supported");
+        }
+
+        let remote_endpoint = match addr {
+            None => None,
+            Some(remote_addr) => Some(remote_addr.try_into()?),
+        };
+
+        if let Some(endpoint) = remote_endpoint {
+            self.try_bind_empheral(&endpoint)?;
+        }
+
+        let mut total_bytes = 0;
+        for io_vec in io_vec_iter {
+            let buffer = {
+                let io_vec = io_vec?;
+                let mut buffer = vec![0u8; io_vec.len()];
+                io_vec.read_from_user(&mut buffer)?;
+                buffer
+            };
+
+            // TODO: Block if the send buffer is full
+            let sent_bytes = self.try_sendto(&buffer, remote_endpoint, flags)?;
+            total_bytes += sent_bytes;
+        }
+
+        Ok(total_bytes)
+    }
+
+    fn recvmsg(&self, msg_hdr: &mut MessageHeader, flags: SendRecvFlags) -> Result<usize> {
```

Comment:
I don't think it's clear which fields in `msg_hdr: &mut MessageHeader` are input and which are output.

Maybe we can have a better API for this function? Is it possible to put output fields into the return type?

Reply:
This API has been refactored to return `(usize, MessageHeader)`